xrpl/utils/tx_parser/utils.py

```diff
@@ -0,0 +1,1052 @@
+        )
+
+    meta = metadata["meta"]
+    if "AffectedNodes" not in meta or not meta["AffectedNodes"]:  # type: ignore
+        raise XRPLMetadataException("Metadata incomplete: No nodes provided.")
+
+
+def normalize_metadata(
+    metadata: Union[METADATA_TYPE, SUBSCRIPTION_METADATA_TYPE],
+) -> METADATA_TYPE:
+    """Formats the transaction metadata into one standard format.
+
+    Args:
+        metadata (Union[METADATA_TYPE, SUBSCRIPTION_METADATA_TYPE]):
+            Transactions metadata.
+
+    Returns:
+        Metadata in standard format.
+    """
+    transaction = metadata["transaction"]
+    assert isinstance(transaction, dict)
+    normalized_metadata = {
+        tx_field_name: tx_field_value
+        for tx_field_name, tx_field_value in transaction.items()
+    }
+
+    meta = metadata["meta"]
+    assert isinstance(meta, dict)
+    normalized_metadata["meta"] = meta  # type: ignore
+    ledger_index = metadata["ledger_index"]
+    assert isinstance(ledger_index, int)
+    normalized_metadata["ledger_index"] = ledger_index
+
+    return normalized_metadata  # type: ignore
+
+
+class FieldName:
+    """FieldName"""
+
+    currency: str
+    issuer: str
+    value: str
+
+
+class FieldState:
+    """FieldState"""
+
+    Balance: Union[AccountBalance, str]
+    LowLimit: Union[AccountBalance, str]
+    HighLimit: Union[AccountBalance, str]
+    Account: str
+    TakerGets: Union[FieldName, str]  # Type: 'NewClass'
+    TakerPays: Union[FieldName, str]  # Type: 'NewClass'
+    Sequence: int
+    Flags: int
+    Expiration: Optional[Union[int, str]] = None
+
+
+class NewFields(FieldState):
+    """NewFields"""
+
+    pass
+
+
+class FinalFields(FieldState):
+    """FinalFields"""
+
+    pass
+
+
+class PreviousFields(FieldState):
+    """PreviousFields"""
+
+    pass
+
+
+class NormalizedNode:
+    """A standard format for nodes."""
+
+    def __init__(
+        self: NormalizedNode,
+        diff_type: str,
+        entry_type: str,
+        ledger_index: str,
+        new_fields: Optional[FieldState],
+        final_fields: Optional[FieldState],
+        previous_fields: Optional[FieldState],
+    ) -> None:
+        """
+        Args:
+            diffType (str): Node type (ModifiedNode, CreatedNode or DeletedNode).
+            entryType (str): Entry type (e.g. Offer, AccountRoot, …).
+            ledgerIndex (str): Ledger index.
+            newFields (Optional[FieldState]):
+                New fields.
+            finalFields (Optional[FieldState]):
+                Fields after the transaction occurred.
+            previousFields (Optional[FieldState]):
+                Fields before the transaction occurred.
+        """
+        self.diff_type = diff_type
+        self.entry_type = entry_type
+        self.ledger_index = ledger_index
+        self.new_fields = new_fields
+        self.final_fields = final_fields
+        self.previous_fields = previous_fields
+
+
+def field_factory(
```

Comment:
This method feels very hacky. Typing doesn't need to be so restrictive in this package that we need to do some metaprogramming in order to use it.

Reply:
I found a way to make this function way easier 👍